main-settings/src/main/scala/sbt/Scope.scala

```diff
@@ -278,20 +279,50 @@ object Scope {
   )(rawScope: Scope): Seq[Scope] = {
     val scope = Scope.replaceThis(GlobalScope)(rawScope)
 
-    def nonProjectScopes(resolvedProj: ResolvedReference)(px: ScopeAxis[ResolvedReference]) = {
-      val p = px.toOption getOrElse resolvedProj
-      val configProj = p match {
-        case pr: ProjectRef => pr; case br: BuildRef => ProjectRef(br.build, rootProject(br.build))
-      }
-      val cLin = scope.config match {
-        case Select(conf) => index.config(configProj, conf); case _ => withZeroAxis(scope.config)
-      }
+    def nonProjectScopes(resolvedProj: ResolvedReference)(
+        pLin: Seq[ScopeAxis[ResolvedReference]]): Seq[Scope] = {
       val tLin = scope.task match {
-        case t @ Select(_) => linearize(t)(taskInherit); case _ => withZeroAxis(scope.task)
+        case t @ Select(k) =>
+          if (taskInherit(k).isEmpty) t :: Zero :: Nil
+          else linearize(t)(taskInherit)
+        case _ => withZeroAxis(scope.task)
       }
       val eLin = withZeroAxis(scope.extra)
-      for (c <- cLin; t <- tLin; e <- eLin) yield Scope(px, c, t, e)
+
+      val res = new ArrayBuffer[Scope](pLin.size * 2 * tLin.size * eLin.size)
+
+      val pLinIt = pLin.iterator
+      while (pLinIt.hasNext) {
+        val px = pLinIt.next()
+        val p = px.toOption getOrElse resolvedProj
+        val configProj = p match {
+          case pr: ProjectRef => pr;
+          case br: BuildRef   => ProjectRef(br.build, rootProject(br.build))
+        }
+
+        val cLin = scope.config match {
+          case Select(conf) => index.config(configProj, conf)
+          case _            => withZeroAxis(scope.config)
+        }
+
+        val cLinIt = cLin.iterator
+        while (cLinIt.hasNext) {
+          val c = cLinIt.next()
+          val tLinIt = tLin.iterator
+          while (tLinIt.hasNext) {
+            val t = tLinIt.next()
+            val eLinIt = eLin.iterator
```

Comment:
We normally don't use extra axis, so in most cases this would contain `List(Zero)`.
Maybe it could checking for size == 1 make it faster?

Reply:
:+1: I guess there's a whole set of potential optimizations still left for optimizing walking the delegation chain. The old code tried to use `toStream` to compute the chain lazily. However, the overhead of `Stream` was much bigger than the gains. I haven't looked into how the delegation chains look exactly in general but I'm pretty sure there a some cleverer ways to walk it lazily.